### PR TITLE
[libcxx] Correct and clean-up filesystem operations error_code paths

### PR DESCRIPTION
**Pull request by 4-rodrigo-salazar**
3 error_code related cleanups/corrections in the std::filesystem operations functions.

1. In `__copy`, the `ec->clear()` is unnecessary as `ErrorHandler` at the start of each function clears the error_code as part of its initialization.

2. In `__copy`, in the recursive codepath we are not checking the error_code result of `it.increment(m_ec2)` immediately after use in the for loop condition (and we aren't checking it after the final increment when we don't enter the loop).

3. In `__weakly_canonical`, it makes calls to `__canonical` (which internally uses OS APIs implementing POSIX `realpath`) and we are not checking the error code result from the `__canonical` call. Both `weakly_canonical` and `canonical` are supposed to set the error_code when underlying OS APIs result in an error (https://eel.is/c++draft/fs.err.report#3.1). With this change we propagate up the error_code from `__canonical` caused by any underlying OS API failure up to the `__weakly_canonical`. Essentially, if `__canonical` thinks an error code should be set, then `__weakly_canonical` must as well. Before this change it would be throwing an exception in the non-error_code form of the function when `__canonical` fails, while not setting the error code in the error_code form of the function (an inconsistency).

Added a little coverage in weakly_canonical.pass.cpp for the error_code forms of the API that was missing. Though I am lacking utilities in libcxx testing to add granular testing of the failure scenarios (like forcing realpath to fail for a given path, as it could if you had something like a flaky remote filesystem).

### CODE REVIEW

File: libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.weakly_canonical/weakly_canonical.pass.cpp [hunk 1]
```diff
@@ -79,6 +79,14 @@ int main(int, char**) {
...
                  TEST_WRITE_CONCATENATED(
                      "Input: ", TC.input.string(), "\nExpected: ", expect.string(), "\nOutput: ", output.string()));
+
+    // Get coverage over the error_code form of the api.
+    std::error_code ec;
+    const fs::path output_c = fs::weakly_canonical(p, ec);
+
+    TEST_REQUIRE(PathEq(output_c, expect),
+                 TEST_WRITE_CONCATENATED(
+                     "Input: ", TC.input.string(), "\nExpected: ", expect.string(), "\nOutput: ", output_c.string()));
```
Comment (ldionne) [inline libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.weakly_canonical/weakly_canonical.pass.cpp:89]: Minor style nit, I'd suggest writing it this way instead:

```suggestion
    // Get coverage over the error_code form of the api.
    {
      std::error_code ec;
      const fs::path output = fs::weakly_canonical(p, ec);

      TEST_REQUIRE(PathEq(output, expect),
                   TEST_WRITE_CONCATENATED(
                       "Input: ", TC.input.string(), "\nExpected: ", expect.string(), "\nOutput: ", output.string()));
    }
```

Similarly, put the exception-based test right above in its own scope. That way it's more obvious that those are two test cases and you don't need to invent a new name (`output_c`) for one of the test cases.

File: libcxx/src/filesystem/operations.cpp [hunk 2]
```diff
@@ -166,15 +163,15 @@ void __copy(const path& from, const path& to, copy_options options, error_code*
       return;
     }
     error_code m_ec2;
-    for (; it != directory_iterator(); it.increment(m_ec2)) {
-      if (m_ec2) {
-        return err.report(m_ec2);
-      }
+    for (; !m_ec2 && it != directory_iterator(); it.increment(m_ec2)) {
```
Comment (ldionne) [inline libcxx/src/filesystem/operations.cpp:166]: I don't understand this change. IIUC, the PR description mentions that we incorrectly forgot to check `m_ec2` after the recursive call to `__copy`, but this didn't change after your patch. We're still passing `ec` to the recursive `__copy` call (not `m_ec2`). I think I am missing something.

**Review by ldionne — COMMENTED** [review]
This seems mostly good to me, but I have a question about something I don't quite understand. Thanks for noticing these issues!

File: libcxx/src/filesystem/operations.cpp (hunk 2 above)
Comment (4-rodrigo-salazar) [reply libcxx/src/filesystem/operations.cpp:166]: It's that we don't check m_ec2 after the call to 'it.increment(m_ec2)'

After the for loops increment we now check m_ec2 as the loop condition, and additionally we make sure now to check the code after the final increment whereas we wouldn't before.

Comment (ldionne) [reply libcxx/src/filesystem/operations.cpp:166]: Ok, I see it now, thanks!

**Review by ldionne — APPROVED** [review]
LGTM, thanks for the fix! I applied the nit comment I had.

**Pull request merged**